contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,286 @@
+
+    mapping(address => uint256) public balances;
+    mapping(bytes32 => Game) public games;
+
+    address private constant NULL_ADDRESS = address(0);
+    string private constant INVALID_PLAYER_MSG = "Invalid player";
+    string private constant GAME_NOT_FOUND_MSG = "Game not found";
+    string private constant INVALID_MOVE_MSG = "Invalid move";
+    string private constant INVALID_STEP_MSG = "Invalid step";
+
+    uint256 constant FORFEIT_WINDOW = 12 hours;
+    uint256 constant FEE_PERCENTAGE = 10;
+
+    enum Steps {
+        INIT,
+        PLAYER_ONE_MOVE,
+        PLAYER_TWO_MOVE,
+        PLAYER_ONE_REVEAL,
+        PLAYER_TWO_REVEAL,
+        END
+    }
+
+    struct Game {
+        uint256 stake;
+        address playerOne;
+        address playerTwo;
+        uint8 playerTwoMoveHash;
+        uint8 playerOneMove;
+        uint8 playerTwoMove;
+        uint256 expiryDate;
+        uint8 step;
+    }
+
+    event PlayerOneMoves(
+        address indexed player,
+        bytes32 indexed gameKey,
+        uint256 value
+    );
+
+    event PlayerTwoMoves(
+        address indexed player,
+        bytes32 indexed gameKey,
+        uint256 value
+    );
+
+    event PlayerTwoReveals(
+        address indexed player,
+        bytes32 indexed gameKey,
+        uint8 move
+    );
+
+    event WinnerRewarded(
+        address indexed player,
+        bytes32 indexed gameKey,
+        uint256 amount
+    );
+
+    event  ForfeitPaid(
+        address indexed player,
+        bytes32 indexed gameKey,
+        uint256 amount
+    );
+
+    event WithDraw(
+        address indexed withdrawer,
+        uint amount
+    );
+
+    // The creator's move is embedded in the game key
+    function playerOneMove(bytes32 gameKey, uint stake) external payable whenNotPaused {
+
+        require(gameKey != bytes32(0), "Game key cannot be empty");
+        require(stake > 0, "Stake cannot be zero");
+
+        Game storage game = games[gameKey];
+        require(game.step == Steps.INIT, INVALID_STEP_MSG);
+
+        game.playerOne = msg.sender;
+        game.stake = stake;
+        game.step = Steps.PLAYER_ONE_MOVE;
+        depositStake(stake);
+
+        emit PlayerOneMoves(msg.sender, gameKey, stake);
+    }
+
+    /*
+    * The opponent joins the game
+    */
+    function playerTwoMove(bytes32 gameKey, bytes32 moveHash) external payable whenNotPaused {
+
+        Game storage game = games[gameKey];
+        uint256 stake = game.stake;
+        require(stake != 0, GAME_NOT_FOUND_MSG);
+        require(game.step == Steps.PLAYER_ONE_MOVE, INVALID_STEP_MSG);
+
+        game.playerTwo = msg.sender;
+        game.playerTwoMoveHash = moveHash;
+        game.step = steps.PLAYER_TWO_MOVE;
+        depositStake(stake);
+
+        emit PlayerTwoMoves(msg.sender, msg.value, gameKey);
+    }
+
+    function depositStake(uint256 stake) internal {
+        uint256 senderBalance = balances[msg.sender];
+
+        if (msg.value == stake) {
+            return;
```

Comment:
Yes, you are trying to save gas, but then what is the point of the `SLOAD` on line 116?

Reply:
Ah. I had spotted this. Sorry. This is quite out of date. I hadn't pushed my changes for a while.